Variant

redboltz opened the pull request:
This PR provides a msgpack::type::variant type and its adaptor. It uses Boost.Variant. We can  treat any msgpack::object with msgpack::type::variant. (convert/pack/create msgpack::object)

msgpack::type::variant_ref is also provided. When we convert to msgpack::type::varinat from msgpack::object, reference types are used.

See the following table:

## In common msgpack::type::variant and msgpack::type::variant:

|msgpack::type |variant type |
|--- |--- |
|NIL  |msgpack::type::nil|
|BOOLEAN  |bool|
|POSITIVE_INTEGER  |uint64_t|
|NEGATIVE_INTEGER  |int64_t|
|FLOAT  |double|

## msgpack::type::variant

|msgpack::type |variant type |
|--- |--- |
|STR  |std::string|
|BIN  |std::vector<char>|
|EXT  |msgpack::type::ext|
|ARRAY  |std::vector<msgpack::type::variant>|
|MAP  |std::multimap<msgpack::type::variant, msgpack::type::variant>|

## msgpack::type::variant_ref

|msgpack::type |variant type |
|--- |--- |
|STR  |boost::string_ref|
|BIN  |msgpack::type::raw_ref|
|EXT  |msgpack::type::ext_ref|
|ARRAY  |std::vector<msgpack::type::variant_ref>|
|MAP  |std::multimap<msgpack::type::variant, msgpack::type::variant_ref>|

jpetso commented on example/boost/msgpack_variant_mapbased.cpp:
```diff
@@ -0,0 +1,100 @@
+            else if (key == "address") {
+                // When you want to append string
+                // "Tokyo" -> "Tokyo, JAPAN"
+                // Use msgpack::type::variant instead of msgpack::type::varinat_ref
```
Typo: should be `msgpack::type::variant_ref`

redboltz replied:
Yes, thanks for the review. I will fix it.

jpetso commented on include/msgpack/adaptor/boost/msgpack_variant.hpp:
```diff
@@ -0,0 +1,391 @@
+typedef boost::make_recursive_variant<
+    nil,               // NIL
+    bool,              // BOOL
+    int64_t,           // NEGATIVE_INTEGER
```
I think it warrants a mention that if you assign a positive int64_t, it will still be encoded as POSITIVE_INTEGER but remain in the variant as the int64_t that was assigned.

redboltz replied:
True. The following rule is applied.

### Packing

|msgpack::type::variant|actual value|msgpack bytes|
|----|---|--------|-----------|
|uint64_t|positive integer|positive fixnum or uint N|
|int64_t|positive integer|positive fixnum or uint N|
|int64_t|negative integer|negative fixnum or int N|

### Unpacking

|msgpack bytes|actual value|msgpack::object::type|
|----|---|--------|
|positive fixnum|positive integer|POSITIVE_INTEGER|
|negative fixnum|negative integer|NEGATIVE_INTEGER|
|int N|positive integer|POSITIVE_INTEGER|
|int N|negative integer|NEGATIVE_INTEGER|
|uint N|positive integer|POSITIVE_INTEGER|

### Converting

|msgpack::object::type|actual value|msgpack::type::variant|
|----|---|--------|
|POSITIVE_INTEGER|positive integer|uint64_t|
|NEGATIVE_INTEGER|positive integer|int64_t|

### Creating object::with_zone

|msgpack::type::variant|actual value|msgpack::object::type|
|----|---|--------|
|int64_t|positive integer|POSITIVE_INTEGER|
|int64_t|negative integer|NEGATIVE_INTEGER|
|uint64_t|positive integer|POSITIVE_INTEGER|

So the comment should be replaced
```C++
    int64_t,           // NEGATIVE_INTEGER
```
with something like
```C++
    int64_t,           // NEGATIVE_INTEGER if actual value < 0 otherwise POSITIVE_INTEGER
```

jpetso commented:
In the interest of ease of use, would it make sense to add convenience accessor methods to the object so I don't have to access the map/multimap with the following:
```
auto* multimap = boost::get<std::multimap<msgpack::type::variant_ref, msgpack::type::variant_ref>>(&variant);
if (multimap) {
    auto iterator = multimap->find("mykey");
}
// Instead, I'd like to do:
if (variant.is_multimap()) {
    auto easy_iterator = variant.multimap().find("mykey");
}
```
Also, can string keys be used like this? If not then that sort of defeats the point of a variant for many use cases. Some example usage should likely be documented too, because the whole boost recursive templating stuff seems confusing even to people who have a rough idea what to expect from boost::variant.

jpetso commented on include/msgpack/adaptor/boost/msgpack_variant.hpp:
```diff
@@ -0,0 +1,391 @@
+namespace detail {
+
+template <typename Stream>
+struct pack_imp:boost::static_visitor<void> {
```
Stylistic opinion: I find it hard to tell the actual type name from the inherited class. Putting spaces around the colon after `pack_imp` would make this a lot more readable, imho.

redboltz replied:
I agree. Updated.

redboltz commented:
Let me clarify the point:

#### Point1

Now, we can write as follows:
```C++
auto* multimap = boost::get<std::multimap<msgpack::type::variant_ref, msgpack::type::variant_ref>>(&variant);
if (multimap) {
    auto iterator = multimap->find("mykey");
}
```

In addition, if the following interface is supported, it's convenient:
```C++
if (variant.is_multimap()) {
    auto easy_iterator = variant.multimap().find("mykey");
}
```

Do I understand correctly?

#### Point2
You want to know that we can write string literal directly as follows:
```C++
    auto iterator = multimap->find("mykey");
```

Unfortunately, we cant. We need to write as follows:
```C++
    auto iterator = multimap->find(std::string("mykey"));
```

In msgpack::type::variant(_ref), boost::variant includes bool and std::string. Character literals are decayed to const char* and const char* is converted to bool. It has higher priority than const char* to std::string conversion.

As the result of that, any keys are converted to true, then find() is matched with any keys.
```C++
    auto iterator = multimap->find("anykey");
```

jpetso commented:
Yes, that is what I was asking. Thank you.

redboltz commented:
Thanks @jpetso.

#### Point1

Maybe I can add some convenient functions. But It's not a member function but a free function. Because msgpack::type::variant is a typedef of boost::variant. I cannot add a member function to that.

#### Point2

I think that it is useful and easy to understand that character literals converted to std::string. As far as msgpack::type::variant is a typedef of boost::variant, I can't do that.



If there is a better way, I want to know and I'm ready to rewrite my PR. I've tried to inherit boost::variant as follows:

```C++
struct variant : boost::variant< .... >
```

My expectation was I can insert some conversion in constructor. But it didn't work.

Also I tried a composition approach.

```C++
struct variant {
    boost::variant< .... >
};
```
But boost::variant has various interfaces. For example static visitor. It's too hard to implement all delegation interfaces.

Perhaps the composition approach with get native implementation approach would be better such as get_boost_variant().

Anyway, my first PR is "use boost::variant directly." 

Review and comments are very welcome :)

jpetso commented:
> My expectation was I can insert some conversion in constructor. But it didn't work.

Can you describe the problem that you ran into with a few extra details? What kind of conversion did you try to implement in the constructor?

redboltz commented:
I don't keep the code. So I wrote the minimal example code to describe the conversion:

http://melpon.org/wandbox/permlink/KbZjhZBk6wphJcHF

And it works as I expected. So I looked over something when I tried before. I will refine the PR to use inheriting approach instead of using boost::variant directly.

redboltz commented:
@jpetso, I updated the PR. Client can use string literals without std::srting(). In addition, I added integer family constructor. We can pass integer literal directly. In the constructor, the argument is checked and if the argument is negative, variant, the base class, is assigned as int64_t, else uint64_t. So int64_t always has a negative value. I think that it is easier to understand.
I also added type checking functions and getting functions. Those are is_* and as_*. I added as_ prefix to getting functions because some of name of types are conflicted with C++ keyword e.g. double.

Could you review my update?
